### DISCUSSION TITLE
Cannot cast the type into its generic parameter using operator (op_Implicit or op_Explicit) even if nothing runs wrong (using reflection)

### DISCUSSION THREAD

**Discussion by CaryCatZ** [dotnet/runtime#94633] [General] [answered]
It seems an operator declaration with a generic parameter of class would not work.
`TestClass<string> as string` does not work even if TestClass defines `operator T(TestClass<T>)`.

An interface:
```C#
namespace TestProject1;

public interface IWrapper<TWrapper, TWrapped> where TWrapper : IWrapper<TWrapper, TWrapped>
{
  public TWrapped Wrapped { get; }

  public static virtual implicit operator TWrapped(TWrapper @this) => @this.Wrapped;
}
```
It represents a wrapper and it can be cast into `TWrapped`.

And two same wrappers:
```C#
namespace TestProject1;

public class Wrapper1<TWrapped> : IWrapper<Wrapper1<TWrapped>, TWrapped>
{
  public Wrapper1(TWrapped wrapped)
  {
    Wrapped = wrapped;
  }

  public TWrapped Wrapped { get; }
}

```
and
```C#
namespace TestProject1;

public class Wrapper2<TWrapped> : IWrapper<Wrapper2<TWrapped>, TWrapped>
{
  public Wrapper2(TWrapped wrapped)
  {
    Wrapped = wrapped;
  }

  public TWrapped Wrapped { get; }
}

```
and
```C#
namespace TestProject1;

public class GenericCast<T>
{
  public T val;

  public GenericCast(T val)
  {
    this.val = val;
  }

  public static implicit operator T(GenericCast<T> @this) => @this.val;
}

```

However, now, the cast from `Wrapper1` to `Wrapper2` does not work.
```C#
var wrapper1 = new Wrapper1<Wrapper2<int>>(new Wrapper2<int>(42));
// Console.WriteLine((wrapper1 as Wrapper2<int>).Wrapped); // does not work, even if `TWrapped` is `Wrapper2<int>`

var g = new GenericCast<string>("Value");
// Console.WriteLine(g as string); // does not work
// Console.WriteLine(GenericCast<string>.op_Implicit(g)); // CS0571
Console.WriteLine(typeof(GenericCast<string>).GetMethod("op_Implicit").Invoke(null, new object?[]{g})); // reflection works well
```

Even though `operator TWrapped` were moved to `Wrapper1` and `Wrapper2`, it would not work, either.
Look at `GenericCast<T>`, although `operator T` is declared, the cast from `GenericCast<string>` to `string` does not work, either.

I think this is strange, is this a bug?

**Comment by saucecontrol** [comment, marked as answer]
From:
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#as-operator

> The `as` operator considers only reference, nullable, boxing, and unboxing conversions. You can't use the `as` operator to perform a user-defined conversion. To do that, use a [cast expression](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#cast-expression).

**Reply by CaryCatZ** [reply]
Thanks a lot, problem solved.
